### PR TITLE
cpu_flag:new cases to test cpu flags

### PR DESCRIPTION
**Pull request by nanliu-r**
ID: 1855513, 1855512

Signed-off-by: Nana Liu <nanliu@redhat.com>

### CODE REVIEW

**Comment by nanliu-r** [comment]
Test results:
 (2/4) Host_RHEL.m8.u2.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.0.x86_64.io-github-autotest-qemu.x86_cpu_flags.intel.avx512_series_flags_2: PASS (64.41 s)
 (3/4) Host_RHEL.m8.u2.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.0.x86_64.io-github-autotest-qemu.x86_cpu_flags.intel.avx512_series_flags_4: PASS (65.18 s)
 (4/4) Host_RHEL.m8.u2.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.0.x86_64.io-github-autotest-qemu.x86_cpu_flags.intel.pt_vmx: PASS (62.52 s)

 (2/4) Host_RHEL.m8.u2.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.0.x86_64.io-github-autotest-qemu.x86_cpu_flags.intel.avx512_series_flags_2.q35: PASS (219.32 s)
 (3/4) Host_RHEL.m8.u2.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.0.x86_64.io-github-autotest-qemu.x86_cpu_flags.intel.avx512_series_flags_4.q35: PASS (63.85 s)
(4/4) Host_RHEL.m8.u2.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.0.x86_64.io-github-autotest-qemu.x86_cpu_flags.intel.pt_vmx.q35: PASS (133.32 s)

 (2/4) Host_RHEL.m8.u2.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.x86_cpu_flags.intel.avx512_series_flags_2.q35: PASS (245.57 s)
 (3/4) Host_RHEL.m8.u2.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.x86_cpu_flags.intel.avx512_series_flags_4.q35: PASS (197.54 s)
 (4/4) Host_RHEL.m8.u2.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.x86_cpu_flags.intel.pt_vmx.q35: PASS (375.48 s)

**Comment by nanliu-r** [comment]
@vivianQizhu @huangyum Please help to review this, thanks a lot.

File: qemu/tests/cpu_flag_intel_pt.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+        session.close()
+        vm.destroy()
+
+    vendor_id = params.get("vendor_id", "")
+    if vendor_id:
+        if vendor_id != cpu.get_vendor():
+            test.cancel("Need host vendor %s to support this test case" % vendor_id)
+
+    flags = params["flags"]
+    check_host_flags = params.get_boolean("check_host_flags")
```
Comment (huangyum) [inline qemu/tests/cpu_flag_intel_pt.py:55]: You can't just check host flags for intel_pt. E.g. On Cascadelake host, host has intel_pt flag, but it doesn't support this feature.

File: qemu/tests/cpu_flag_intel_pt.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+def run(test, params, env):
+    """
+    Test cpu flag intel-pt.
+    1) Check if current flags are in the supported lists, if no, cancel test
+    2) Otherwise, set pt_model = 1 on host at first
+    3) For q35
+       3.1) boot guest with cpu model 'host' without intel-pt
+       3.2) boot guest with cpu model 'Icelake-Server' with intel-pt
+    4) For pc
+       4.1) boot guest with intel-pt and min-level=0x14
```
Comment (huangyum) [inline qemu/tests/cpu_flag_intel_pt.py:18]: Would you please have a try with '-cpu host -M pc'? Besides,  maybe step 3 and step 4 can be implemented in cfg?

File: qemu/tests/cpu_flag_intel_pt.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    5) Check cpu flags in guest(only for linux guest)
+    6) Reboot guest
+    7) Set pt_model = 0 on host at last
```
Comment (huangyum) [inline qemu/tests/cpu_flag_intel_pt.py:21]: Had better get pt_mode before test and restore it after test.

File: qemu/tests/cpu_flag_intel_pt.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+
+@error_context.context_aware
+def run(test, params, env):
+    """
+    Test cpu flag intel-pt.
+    1) Check if current flags are in the supported lists, if no, cancel test
+    2) Otherwise, set pt_model = 1 on host at first
+    3) For q35
+       3.1) boot guest with cpu model 'host' without intel-pt
+       3.2) boot guest with cpu model 'Icelake-Server' with intel-pt
```
Comment (huangyum) [inline qemu/tests/cpu_flag_intel_pt.py:16]: I guess you can use auto cpu model for this scenario,

File: qemu/tests/cpu_flag_intel_pt.py [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+            test.cancel("Need host vendor %s to support this test case" % vendor_id)
+
+    flags = params["flags"]
+    check_host_flags = params.get_boolean("check_host_flags")
+    if check_host_flags:
+        check_flags(params, flags, test)
+
+    try:
+        process.system('rmmod kvm_intel')
+        process.system('modprobe kvm_intel pt_mode=1')
```
Comment (huangyum) [inline qemu/tests/cpu_flag_intel_pt.py:61]: Prefer to put these commands in cfg.

File: qemu/tests/cpu_flag_intel_pt.py (hunk 1 above)
Comment (nanliu-r) [reply qemu/tests/cpu_flag_intel_pt.py:55]: > You can't just check host flags for intel_pt. E.g. On Cascadelake host, host has intel_pt flag, but it doesn't support this feature.

How about we pick up the specific host to test the "intel-pt"? It is not flexible very much but very effective.

File: qemu/tests/cpu_flag_intel_pt.py (hunk 4 above)
Comment (nanliu-r) [reply qemu/tests/cpu_flag_intel_pt.py:16]: > I guess you can use auto cpu model for this scenario,

Is there possible that the auto model is duplicate with 'host'?

File: qemu/tests/cpu_flag_intel_pt.py (hunk 1 above)
Comment (huangyum) [reply qemu/tests/cpu_flag_intel_pt.py:55]: Did you get any advice from developer?

File: qemu/tests/cpu_flag_intel_pt.py (hunk 4 above)
Comment (huangyum) [reply qemu/tests/cpu_flag_intel_pt.py:16]: No, the auto selected models are listed in cpu.py in vt, which doesn't include 'host'.

**Comment by vivianQizhu** [comment]
@nanliu-r Do we have any conclusion for this one?

File: qemu/tests/cpu_flag_intel_pt.py (hunk 1 above)
Comment (nanliu-r) [reply qemu/tests/cpu_flag_intel_pt.py:55]: > Did you get any advice from developer?

not yet.

**Comment by nanliu-r** [comment]
> @nanliu-r Do we have any conclusion for this one?
Seems I need to delay one case of them since flag 'intel-pt' needs more investigation.Will update it later.

File: qemu/tests/cpu_flag_intel_pt.py (hunk 4 above)
Comment (nanliu-r) [reply qemu/tests/cpu_flag_intel_pt.py]: Thanks, will separate one new patch for the flag `intel-pt`  since it needs more time to investigate.

**Comment by nanliu-r** [comment]
will update test results once I get machine.

**Comment by nanliu-r** [comment]
(3/8) Host_RHEL.m8.u3.product_rhel.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.9.x86_64.io-github-autotest-qemu.x86_cpu_flags.intel.avx512_series_flags_2: PASS (84.50 s)
 (4/8) Host_RHEL.m8.u3.product_rhel.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.9.x86_64.io-github-autotest-qemu.x86_cpu_flags.intel.avx512_series_flags_4: PASS (59.21 s)
 (5/8) Host_RHEL.m8.u3.product_rhel.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.0.x86_64.io-github-autotest-qemu.x86_cpu_flags.intel.avx512_series_flags_2: PASS (65.42 s)
 (6/8) Host_RHEL.m8.u3.product_rhel.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.0.x86_64.io-github-autotest-qemu.x86_cpu_flags.intel.avx512_series_flags_4: PASS (63.42 s)
 (7/8) Host_RHEL.m8.u3.product_rhel.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.x86_cpu_flags.intel.avx512_series_flags_2: PASS (465.79 s)
 (8/8) Host_RHEL.m8.u3.product_rhel.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.x86_cpu_flags.intel.avx512_series_flags_4: PASS (183.19 s)

**Comment by nanliu-r** [comment]
@huangyum @vivianQizhu Hi，can you re-review this patch again? thanks a lot.

**Review by vivianQizhu — APPROVED** [review]

**Review by huangyum — APPROVED** [review]
ACK

**Pull request merged**